R/stats-kmeans-tidiers.R

```diff
@@ -42,7 +42,10 @@ tidy.kmeans <- function(x, col.names = colnames(x$centers), ...) {
 #' @export
 #' @seealso [augment()], [stats::kmeans()]
 #' @family kmeans tidiers
-augment.kmeans <- function(x, data, ...) {
+augment.kmeans <- function(x, data = NULL, ...) {
+  if (is.null(data)) {
```

Comment:
In cases like this, where the argument is required, it's actually better to not have a default argument so the user gets an error and knows they have to specify the argument.

Reply:
I'm confused by your comment here.  According to https://github.com/tidymodels/broom/issues/450 

"A number of tidiers do not yet pass the strict tests. Beginning with the changes in #449, the strict = TRUE is the default in tests, so these tidiers can be identified based on the appearance of strict = FALSE in the tests file."

Thus, I thought the overall request was to make revisions such that the tests could run with strict=TRUE.  i.e. `check_arguments(broom:::augment.kmeans, strict = TRUE)`, which currently throws the error message `Arguments data to `augment.kmeans` must have default values. Use `NULL` to indicate missingness if necessary.`. Suggesting that the default argument here should be NULL rather than not having a default..

Also, isn't it better to be explicit here and throw an error saying that data is a required argument  (i.e. `stop("Must specify `data` argument.", call. = FALSE)`)
rather than letting whatever error will surface downstream get fed back to the user?

Reply:
In this particular case we want the default error for consistency. The `kmeans` tidiers are a special case: the `data` is required, whereas with other tidiers it's optional. So if the user doesn't specify `data` they should get the standard R warning for: please give this function that argument it needs that you forgot.

Reply:
Ok, thanks for the explanation.  To close out this discussion, this means the end user message will be:

`Error in rownames(x) : argument "data" is missing, with no default`

I do think there is an opportunity to make the error message more user-friendly such as:

`Must specify "data" argument for augment.kmeans.`

but will defer to you and remove my change and add a comment so future developers are aware not to attempt a similar change.

For convenience, here's a reprex for the current error message.

```
set.seed(2)
x <- rbind(
  matrix(rnorm(100, sd = 0.3), ncol = 2),
  matrix(rnorm(100, mean = 1, sd = 0.3), ncol = 2)
)
fit <- kmeans(x, 2)
augment(fit)
## Error in rownames(x) : argument "data" is missing, with no default 
```

Reply:
If you feel strongly that the original error message is better I'm happy to go with that!

Reply:
Cool, thanks for the flexibility!  I can still remember the days before I knew how to dig into R source code and how unfriendly some R error messages can be.

Reply:
Just pushed an update that updates my original error message from `Must specify "data" argument.` to `Must specify "data" argument for augment.kmeans.`

Resolving conversation.